File: InterviewBit/heaps-and-maps/lru-cache.py
class LinkedListNode:
    def __init__(self, value):
        self.value = value
        self.prev = None
        self.next = None


class LinkedList:
    def __init__(self):
        self.count = 0
        self.head = None
        self.tail = None

    def append(self, value):
        node = LinkedListNode(value)
        if self.tail:
            node.prev = self.tail
            self.tail.next = node
            self.tail = self.tail.next
        else:
            self.head = self.tail = node
        self.count += 1
        return node

    def prepend(self, value):
        node = LinkedListNode(value)
        if self.head:
            node.next = self.head
            self.head.prev = node
            self.head = self.head.prev
        else:
            self.head = self.tail = node
        self.count += 1
        return node

    def remove(self, node):
        if node == self.head:
            if self.head.next:
                self.head = self.head.next
                self.head.prev = None
                node.next = None
            else:
                self.head = self.tail = None
        elif node == self.tail:
            if self.tail.prev:
                self.tail = self.tail.prev
                self.tail.next = None
                node.prev = None
            else:
                self.head = self.tail = None
        else:
            prev = node.prev
            next = node.next
            prev.next = next
            next.prev = prev
            node.prev = None
            node.next = None
        self.count -= 1

    def __repr__(self):
        head = self.head
        desc = ""
        while head:
            if len(desc) > 0:
                desc += " -> "
            desc += str(head.value)
            head = head.next
        return desc


class LRUCache:
    # @param capacity, an integer
    def __init__(self, capacity):
        self.capacity = capacity
        self.key_to_list_node = {}
        self.key_value_list = LinkedList()

    def move_node_to_last(self, node):
        key, value = node.value
        self.key_value_list.remove(node)
        new_node = self.key_value_list.append((key, value))
        self.key_to_list_node[key] = new_node

    # @param key, an integer
    # @return an integer
    def get(self, key):
        if key in self.key_to_list_node:
            node = self.key_to_list_node[key]
            k, v = node.value
            self.move_node_to_last(node)
            return v
        else:
            return -1

    # @param key, an integer
    # @param value, an integer
    # @return nothing
    def set(self, key, value):
        if key in self.key_to_list_node:
            node = self.key_to_list_node[key]
            node.value = (key, value)
            self.move_node_to_last(node)
        else:
            node = self.key_value_list.append((key, value))
            self.key_to_list_node[key] = node

        if self.key_value_list.count > self.capacity:
            node = self.key_value_list.head
            k, v = node.value
            if k in self.key_to_list_node:
                del self.key_to_list_node[k]
            self.key_value_list.remove(node)

    def __repr__(self):
        return self.key_value_list.__repr__()


cache = LRUCache(5)
cache.set('A', 1)
print(cache)
cache.set('B', 2)
print(cache)
cache.set('C', 3)
print(cache)
cache.set('D', 4)
print(cache)
cache.set('E', 5)
print(cache)
cache.set('F', 6)
print(cache)
cache.set('G', 7)
print(cache)
cache.set('C', 9)
print(cache)

File: InterviewBit/heaps-and-maps/lru-cache.cpp
#include <iostream>
#include <map>

using namespace std;

template <typename T>
class LinkedListNode
{
public:
    T value;
    LinkedListNode *prev;
    LinkedListNode *next;

    LinkedListNode(T value)
    {
        this->value = value;
        this->prev = NULL;
        this->next = NULL;
    }
};

template <typename T>
class LinkedList
{
private:
    int count;
    LinkedListNode<T> *head;
    LinkedListNode<T> *tail;

public:
    LinkedList()
    {
        this->count = 0;
        this->head = NULL;
        this->tail = NULL;
    }

    LinkedListNode<T> *append(T value)
    {
        LinkedListNode<T> *node = new LinkedListNode<T>(value);
        if (this->tail)
        {
            node->prev = this->tail;
            this->tail->next = node;
            this->tail = this->tail->next;
        }
        else
        {
            this->head = this->tail = node;
        }
        this->count += 1;
        return node;
    }

    LinkedListNode<T> *prepend(T value)
    {
        LinkedListNode<T> *node = LinkedListNode<T>(value);
        if (this->head)
        {
            node->next = this->head;
            this->head->prev = node;
            this->head = this->head->prev;
        }
        else
        {
            this->head = this->tail = node;
        }
        this->count += 1;
        return node;
    }

    void remove(LinkedListNode<T> *node)
    {
        if (this->head == NULL || node == NULL)
        {
            return;
        }

        if (node == this->head)
        {
            if (this->head->next)
            {
                this->head = this->head->next;
                this->head->prev = NULL;
                node->next = NULL;
            }
            else
            {
                this->head = this->tail = NULL;
            }
        }
        else if (node == this->tail)
        {
            if (this->tail->prev)
            {
                this->tail = this->tail->prev;
                this->tail->next = NULL;
                node->prev = NULL;
            }
            else
            {
                this->head = this->tail = NULL;
            }
        }
        else
        {
            LinkedListNode<T> *prev = node->prev;
            LinkedListNode<T> *next = node->next;
            prev->next = next;
            next->prev = prev;
            node->prev = NULL;
            node->next = NULL;
        }
        delete (node);
        this->count -= 1;
    }

    int getCount()
    {
        return this->count;
    }

    LinkedListNode<T> *getHead()
    {
        return this->head;
    }

    LinkedListNode<T> *getTail()
    {
        return this->tail;
    }
};

struct KeyValuePair
{
    int key;
    int value;

    KeyValuePair(int key = -1, int value = -1)
    {
        this->key = key;
        this->value = value;
    }
};

class LRUCache
{
private:
    int capacity;
    LinkedList<KeyValuePair> *valueList;
    map<int, LinkedListNode<KeyValuePair> *> keyToListNode;

    void moveNodeToLast(LinkedListNode<KeyValuePair> *node);

public:
    LRUCache(int capacity);
    int get(int key);
    void set(int key, int value);
};

LRUCache::LRUCache(int capacity)
{
    this->capacity = capacity;
    this->valueList = new LinkedList<KeyValuePair>();
    this->keyToListNode.clear();
}

void LRUCache::moveNodeToLast(LinkedListNode<KeyValuePair> *node)
{
    KeyValuePair kv = node->value;
    this->valueList->remove(node);
    LinkedListNode<KeyValuePair> *newNode = this->valueList->append(kv);
    this->keyToListNode[kv.key] = newNode;
}

int LRUCache::get(int key)
{
    LinkedListNode<KeyValuePair> *node = this->keyToListNode[key];
    if (node)
    {
        KeyValuePair kv = node->value;
        this->moveNodeToLast(node);
        return kv.value;
    }
    else
    {
        return -1;
    }
}

void LRUCache::set(int key, int value)
{
    LinkedListNode<KeyValuePair> *node = this->keyToListNode[key];
    KeyValuePair newKV = KeyValuePair(key, value);
    if (node)
    {
        node->value = newKV;
        this->moveNodeToLast(node);
    }
    else
    {
        node = this->valueList->append(newKV);
        this->keyToListNode[newKV.key] = node;
    }

    if (this->valueList->getCount() > this->capacity)
    {
        LinkedListNode<KeyValuePair> *node = this->valueList->getHead();
        KeyValuePair kvToRemove = node->value;
        this->keyToListNode[kvToRemove.key] = NULL;
        this->valueList->remove(node);
    }
}

int main(int argc, char const *argv[])
{
    LRUCache *cache = new LRUCache(1);
    cache->set(2, 1);
    cache->set(2, 2);
    cout << cache->get(2) << endl;
    cache->set(1, 1);
    cache->set(4, 1);
    cout << cache->get(2) << endl;
    return 0;
}
